apps/homescreen/js/firstrun.js

```diff
@@ -5,10 +5,120 @@
+      (e) => {
+        console.warn('Loading default settings, import failed:', e);
+        return loadDefault();
+      });
+  }
+
+  function importVerticalHome() {
+    return new Promise((resolve, reject) => {
+      if (!navigator.getDataStores) {
+        reject('DataStore API is unavailable');
+        return;
+      }
+
+      navigator.getDataStores(VERTICALHOME_APPS).then(
+        stores => {
+          if (stores.length < 1) {
+            reject('verticalhome datastores inaccessible');
+            return;
+          }
+
+          stores[0].getLength().then(length => {
+            if (length === 0) {
+              reject('No verticalhome app order to import');
+              return;
+            }
+
+            var cursor = stores[0].sync();
+
+            var items = {};
+            var iterateItems = () => {
+              cursor.next().then(task => {
+                switch (task.operation) {
+                  case 'done':
+                    // Convert from verticalhome storage format to something
+                    // homescreen can use, skipping over features that aren't
+                    // supported.
+                    var id;
+                    var order = [];
+                    for (var prop in items) {
+                      var item = items[prop];
+                      switch (item.type) {
+                        case 'app':
+                          id = item.manifestURL + '/' +
+                            (item.entryPoint ? item.entryPoint : '');
+                          break;
+
+                        case 'bookmark':
+                          id = item.id;
+                          break;
+
+                        default:
+                          continue;
+                      }
+
+                      order.push({ id: id, order: item.index });
+                    }
+
+                    // Sort by order
+                    order.sort((a, b) => { return a.order - b.order; });
+
+                    // Rewrite order so there are no gaps
+                    order.forEach((entry, index) => {
+                      entry.order = index;
+                    });
+
+                    // Load columns setting and resolve
+                    navigator.getDataStores(VERTICALHOME_SETTINGS).then(
+                      stores => {
+                        stores[0].get(COLS_PREF).then(cols => {
+                          resolve({ order: order, small: cols >= 4 });
+                        }, () => {
+                          resolve({ order: order, small: false });
+                        });
+                      },
+                      (e) => {
+                        console.error('Error retrieving columns preference', e);
+                        resolve({ order: order, small: false });
+                      });
+                    return;
+
+                  case 'clear':
+                    items = {};
+                    break;
+
+                  case 'put':
```

Comment:
verticalhome never calls remove, but I can see about adding it anyway (it just clears and re-saves everything because the indexes change - it only calls put when the cached icon changes).